[WIP, 1.9] Bootstrapping BSD Sockets like API

pfalcon opened the pull request:
This shows the very initial steps on moving BSD Sockets API (prototyped out of tree) into the tree.

Things to decide/ack:

* Location. While later we may need tighter integration with subsys/net/ip/, so far everything fits nicely into subsys/net/lib/sockets/.
* Prefix to use for functions. As was discussed during RFC, there should be prefix, and optionally a header should #define pure POSIX names.
* Header name/location.
* Changes to net_context and friends. As can be seen, the only addition is a per-socket packet/connection queue. There's a recast of existing bit, and fairly speaking, I need to track more detailed context status, still looking how to do that, if anything there're whole 4 bytes of user_data pointer.

pfalcon committed:
net: sockets: Bootstrap Sockets API implementation

This adds Kconfig and build infrastructure and implements
zsock_socket() and zsock_close() functions.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon committed:
net: context, pkt: Changes for Sockets API

Two changes are required so far:

* There's unavoidable need to have a per-socket queue of packets
(for data sockets) or pending connections (for listening sockets).
These queues share the same space (as a C union).
* There's a need to track "EOF" status of connection, synchronized
with a queue of pending packets (i.e. EOF status should be processed
only when all pending packets are processed). A natural place to
store it per-packet then, and we had a "sent" bit which was used
only for outgoing packets, recast it as "eof" for incoming socket
packets.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon commented:
@jukkar , @tbursztyka , @Likailiu, @simple2017 , @mike-scott : Early comments are welcome.

jukkar commented on subsys/net/lib/Kconfig:
```diff
@@ -6,6 +6,8 @@
 
 menu "Network Protocols"
 
+source "subsys/net/lib/sockets/Kconfig"
```
I am fine with having a separate directory for this so we do not clutter subsys/net/ip too much.

jukkar commented on subsys/net/lib/sockets/sockets.c:
```diff
@@ -0,0 +1,20 @@
+	SET_ERRNO(net_context_get(family, type, proto, &ctx));
+
+	/* TODO: Ensure non-negative */
+	return (int)ctx;
```
Better to use POINTER_TO_INT(ctx) here.

pfalcon replied:
Fixed.

jukkar reviewed (commented):
For the commit subject, I would prefer more logical naming like "net: socket: ...."
I usually try to avoid putting directory or file name into commit subject as files are moved around etc so better imho is to use logical component names here.

pfalcon commented:
> For the commit subject, I would prefer more logical naming like "net: socket: ...."

Fixed, thanks.

pfalcon committed:
net: sockets: Add configurable option to provide raw POSIX API names

With CONFIG_NET_SOCKETS_POSIX_NAMES=y, "raw" POSIX names like
socket(), recv(), close() will be exposed (using macro defines).
The close() is the biggest culprit here, because in POSIX it
applies to any file descriptor, but in this implementation -
only to sockets.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon committed:
net: context: Allow to put context into FIFO at expense of user_data

By moving user_data member at the beginning of structure. With
refcount at the beginning, reliable passsing of contexts via
FIFO was just impossible. (Queuing contexts to a FIFO is required
for BSD Sockets API).

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

nashif commented on subsys/net/lib/sockets/Kconfig:
```diff
@@ -0,0 +1,28 @@
+        with standard POSIX names like socket(), recv(), and close(), to help
+        with porting existing code. Note that close() may require a special
+        attention, as in POSIX it closes any file descriptor, while with this
+        option enaled, it will still apply only to sockets.
```
can you give an example where such namespace clashes might occur?

pfalcon replied:
Sure, Zephyr is being used together with a 3rd-party library, whose authors also doubted that namespace clashes would occur, so used names like "listen" and "close" in their API ;-).

nashif replied:
Can you give a specific example of such 3rd-party library? Not sure how we ever have 2 providers of bsd socket APIs, What libraries that potentially will be used with Zephyr would have implemented those APIs? This is in most cases implemented by the OS, not libraries.

pfalcon replied:
An example is lwIP. Another example is someone finally addressing one of the big functionality holes of Zephyr - lack of native POSIX port.

> Not sure how we ever have 2 providers of bsd socket APIs, What libraries that potentially will be used with Zephyr would have implemented those APIs?

I'm not sure either, and that's good indication that best practices of namespacing should be followed, otherwise we get a situation of a crowd where everyone considers oneself exceptional and that problems of everyone else doesn't apply to them, and if they happen, it's somebody else's chore to resolve them.

> This is in most cases implemented by the OS, not libraries.

Zephyr is itself a library (libzephyr.a) which real applications (consisting of many other libraries) use to get some hardware abstraction and multitasking. Maybe in a couple of decades it'll become "the OS", but standing in a big-big crowd and pretending it's the only thing isn't the smoothest way to get there ;-).

Let's again take lwIP as an example. It defines 2 related config params:

~~~
#if LWIP_COMPAT_SOCKETS
#define accept(a,b,c)         lwip_accept(a,b,c)
#define bind(a,b,c)           lwip_bind(a,b,c)
...
#if LWIP_POSIX_SOCKETS_IO_NAMES
#define read(a,b,c)           lwip_read(a,b,c)
#define write(a,b,c)          lwip_write(a,b,c)
#define close(s)              lwip_close(s)
...
#endif
#endif
~~~

As the paragraph on which you comment describes, close() alone is enough to decide - it's irresponsible to grab that name for a code which doesn't do all that normal close() does, then it should be prefixed, then everything else should be prefixed for consistency.

pfalcon committed:
WIP: kernel: fifo: Add peek_head/peek_tail accessors.

TODO: Should really go to kernel.h.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon committed:
net: sockets: Implement bind(), connect(), listen(), accept()

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon committed:
net: sockets: Add POSIX compat defines for inet_ntop, inet_pton

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon committed:
samples: net: Add socket-based echo server example

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon committed:
net: sockets: Implement send()

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon committed:
net: sockets: Implement recv() for STREAM sockets

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon committed:
net: sockets: Add debug logging config

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon committed:
net: sockets: Explicitly flush conn/pkt queue on close()

If a socket is closed without reading all data from peer or accepting
all pending connection, they will be leaked. So, flush queues
explicitly.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon commented:
Closing in favor of #498, which is resubmit of the complete patchset suitable for wide review. Hopefully, initial comments here were address and/or answered.
